src/textual/app.py

```diff
@@ -1419,14 +1419,36 @@ async def _on_remove(self, event: events.Remove) -> None:
         widget = event.widget
         parent = widget.parent
 
-        widget.reset_focus()
-
-        remove_widgets = widget.walk_children(
-            Widget, with_self=True, method="depth", reverse=True
+        # Get the list of all the widgets that are about to be removed.
+        remove_widgets = list(
```

Comment:
Note I've note made the suggested change to `walk_children` here (have it return a `list`) -- I didn't want to head off into a side-quest while working on this. If we're happy with this particular PR I'll then go tweak `walk_children` and come back and tidy this up.